plugins/soundsourcem4a/soundsourcem4a.cpp

```diff
@@ -1,221 +1,461 @@
+    DEBUG_ASSERT(isValidFrameIndex(frameIndex));
+
+    if (m_curFrameIndex != frameIndex) {
+        MP4SampleId sampleBlockId = kSampleBlockIdMin
+                + (frameIndex / kFramesPerSampleBlock);
+        DEBUG_ASSERT(isValidSampleBlockId(sampleBlockId));
+        if ((frameIndex < m_curFrameIndex) || // seeking backwards?
+                !isValidSampleBlockId(m_curSampleBlockId) || // invalid seek position?
+                (sampleBlockId
+                        > (m_curSampleBlockId + kNumberOfPrefetchSampleBlocks))) { // jumping forward?
+            // Restart decoding one or more blocks of samples backwards
+            // from the calculated starting block to avoid audible glitches.
+            // Implementation note: The type MP4SampleId is unsigned so we
+            // need to be careful when subtracting!
+            if ((kSampleBlockIdMin + kNumberOfPrefetchSampleBlocks)
+                    < sampleBlockId) {
+                sampleBlockId -= kNumberOfPrefetchSampleBlocks;
+            } else {
+                sampleBlockId = kSampleBlockIdMin;
+            }
+            restartDecoding(sampleBlockId);
+            DEBUG_ASSERT(m_curSampleBlockId == sampleBlockId);
+        }
+        // decoding starts before the actual target position
+        DEBUG_ASSERT(m_curFrameIndex <= frameIndex);
+        // prefetch (decode and discard) all samples up to the target position
+        const SINT prefetchFrameCount = frameIndex - m_curFrameIndex;
+        const SINT skipFrameCount =
+                skipSampleFrames(prefetchFrameCount);
+        DEBUG_ASSERT(skipFrameCount <= prefetchFrameCount);
+        if (skipFrameCount != prefetchFrameCount) {
+            qWarning() << "Failed to skip over prefetched sample frames after seeking @" << m_curFrameIndex;
+            // Abort
+            return m_curFrameIndex;
+        }
+    }
+    DEBUG_ASSERT(m_curFrameIndex == frameIndex);
+    return m_curFrameIndex;
 }
 
-Result SoundSourceM4A::parseHeader(){
-    TagLib::MP4::File f(getFilename().toLocal8Bit().constData());
+SINT SoundSourceM4A::readSampleFrames(
+        SINT numberOfFrames, CSAMPLE* sampleBuffer) {
+    DEBUG_ASSERT(isValidFrameIndex(m_curFrameIndex));
+
+    const SINT numberOfFramesTotal = math_min(numberOfFrames,
+            SINT(getFrameIndexMax() - m_curFrameIndex));
+
+    CSAMPLE* pSampleBuffer = sampleBuffer;
+    SINT numberOfFramesRemaining = numberOfFramesTotal;
+    while (0 < numberOfFramesRemaining) {
+        DEBUG_ASSERT(m_sampleBufferReadOffset <=
+                m_sampleBufferWriteOffset);
+        if (m_sampleBufferReadOffset < m_sampleBufferWriteOffset) {
+            // Consume previously decoded sample data
+            const SINT numberOfSamplesDecoded =
+                    m_sampleBufferWriteOffset -
+                    m_sampleBufferReadOffset;
+            const SINT numberOfFramesDecoded =
+                    samples2frames(numberOfSamplesDecoded);
```

Comment:
all the samples2frames to frame conversion here and below makes the code hard to read. Can we just use samples inside this function?